README.md
libyui-mga-gtk
=============

libyui-gtk plugins/extensions for the Mageia tools


ChangeLog
-------------------------------------------------------------------
Sun Jul 11 19:10:53 CEST 2021 - anaselli@linux.it
- New YMGA_CBTable implementation
- built against new libyui tree.
- 1.2.0

-------------------------------------------------------------------
Mon Jun 29 19:19:01 CEST 2020 - anaselli@linux.it
- Added menubar and new menu items implementation
- 1.1.0

-------------------------------------------------------------------
Tue Apr 14 12:12:36 CEST 2020 - anaselli@linux.it
- removed deprecated function
- 1.0.4

-------------------------------------------------------------------
Sun Jan 28 15:59:35 CET 2018 - anaselli@linux.it
- fixed cellChanged (https://bugzilla.redhat.com/show_bug.cgi?id=1524358)
- 1.0.3



src/YGWE.h
/*
  Copyright 2013 by Angelo Naselli <anaselli at linux dot it>
 
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YGWE.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YGWE_h
#define YGWE_h
#include <YExternalWidgets.h>

class YGWE : public YExternalWidgets
{
public:
  YGWE( const std::string& name );
  virtual ~YGWE();

protected:

  virtual YExternalWidgetFactory * createExternalWidgetFactory();
};

#endif //YGWE_h

src/YGWE.cc
/*
  Copyright 2013 by Angelo Naselli <anaselli at linux dot it>
 
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YGWE.cc

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#include "YGWE.h"
#include "YMGAGWidgetFactory.h"
#include <YUIException.h>
#include <string.h>

YGWE::YGWE( const std::string& name ) : YExternalWidgets(name)
{
}

YGWE::~YGWE()
{
}

YExternalWidgetFactory* YGWE::createExternalWidgetFactory()
{
  YMGAGWidgetFactory * factory = new YMGAGWidgetFactory();
  YUI_CHECK_NEW( factory );

  return factory;
}

YExternalWidgets * createExternalWidgets(const char *name)
{
  std::string pluginName(name);
  YGWE * ew = new YGWE( pluginName );
  YUI_CHECK_NEW ( ew );

  return YExternalWidgets::externalWidgets(pluginName);
}


src/YMGAGMenuBar.h
/*
  Copyright 2020 by Angelo Naselli <anaselli at linux dot it>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/

/*-/

  File:         YMGAGMenuBar.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YMGAGMenuBar_h
#define YMGAGMenuBar_h


#include <yui/mga/YMGAMenuBar.h>
#include <yui/gtk/ygtktreeview.h>

#include <yui/gtk/YGSelectionStore.h>
#include <yui/gtk/YGWidget.h>

#include <gtk/gtk.h>



class YMGAGMenuBar : public YMGAMenuBar, public YGWidget
{
public:

  YMGAGMenuBar ( YWidget *parent );

  virtual ~YMGAGMenuBar ( );

  /**
  * Add an YMenuItem first item represents the menu name, other sub items menu entries
  *
  * Reimplemented from YSelectionWidget.
  **/
  virtual void addItem( YItem * item );

  /**
  * Enable YMGAMenuItem (menu name or menu entry) to enable/disable it into menubar or menu
  *
  * Reimplemented from YMGAMenuBar.
  **/
  virtual void enableItem(YItem * menu_item, bool enable=true);

  /**
  * Hide YMGAMenuItem (menu name or menu entry) to hide/show it into menubar or menu
  *
  * Reimplemented from YMGAMenuBar.
  **/
  virtual void hideItem(YItem * menu_item, bool invisible=true);

  /**
   * Delete all items.
   *
   * Reimplemented from YSelectionWidget
   **/
   virtual void deleteAllItems();

  YGWIDGET_IMPL_COMMON (YMGAMenuBar)


private:
  struct Private;
  Private *d;

  void doCreateMenu (GtkWidget *menu, YItemIterator begin, YItemIterator end);
};

#endif // YMGAGMenuBar_h


src/YMGAGMenuBar.cc
/*
  Copyright 2020 by Angelo Naselli <anaselli at linux dot it>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGAGMenuBar.cc

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/
#define YUILogComponent "mga-gtk-ui"
#include <yui/YUILog.h>

#include <yui/gtk/YGi18n.h>


// #include <yui/gtk/YGUI.h>
#include <yui/gtk/YGUtils.h>
#include <yui/gtk/YGWidget.h>
#include <YSelectionWidget.h>
#include <yui/gtk/YGSelectionStore.h>
#include <yui/gtk/ygtktreeview.h>
#include <yui/gtk/ygtkratiobox.h>
#include <string.h>
#include <yui/mga/YMGAMenuItem.h>
#include <YTable.h>

#include <yui/gtk/YGDialog.h>
#include <gdk/gdkkeysyms.h>
#include <gtk/gtkmenubar.h>
#include <gtk/gtk.h>
#include "YMGAGMenuBar.h"
#include <boost/filesystem.hpp>

#define YGTK_VBOX_NEW(arg) gtk_box_new(GTK_ORIENTATION_VERTICAL,arg)
#define YGTK_HBOX_NEW(arg) gtk_box_new(GTK_ORIENTATION_HORIZONTAL,arg)

typedef std::map<YItem*, GtkWidget*> MenuEntryMap;
typedef std::pair<YItem*, GtkWidget*> MenuEntryPair;

typedef std::map<YItem*, gulong> MenuCBMap;
typedef std::pair<YItem*, gulong> MenuCBPair;


struct YMGAGMenuBar::Private
{
  GtkWidget *menubar;

  MenuEntryMap menu_entry;
  MenuCBMap menu_cb;

  std::vector<GtkWidget*> menu_widgets;
};

YMGAGMenuBar::YMGAGMenuBar(YWidget* parent)
  :  YMGAMenuBar(NULL),
  YGWidget(this, parent, gtk_menu_bar_get_type(), NULL),
  d(new Private)

{
  YUI_CHECK_NEW ( d );
  d->menubar = getWidget();


  //gtk_misc_set_alignment (GTK_MISC (getWidget()), 0.0, 0.5);
}

static void selected_menuitem(GtkMenuItem *, YItem *item)
{
  YGUI::ui()->sendEvent (new YMenuEvent (item));
}

void YMGAGMenuBar::doCreateMenu (GtkWidget *menu, YItemIterator begin, YItemIterator end)
{
  for (YItemIterator it = begin; it != end; it++)
  {
    YMenuSeparator *separator = dynamic_cast<YMenuSeparator *>(*it);
    if (separator)
    {
      GtkWidget *sep = gtk_separator_menu_item_new();
      gtk_menu_shell_append(GTK_MENU_SHELL(menu), sep);
      gtk_widget_show(sep);
    }
    else
    {
      GtkWidget *entry;
      YItem * yitem = *it;
      std::string action_name = YGUtils::mapKBAccel (yitem->label());
      if (yitem->hasIconName())
      {
        GtkIconTheme * theme = gtk_icon_theme_get_default();
        std::string ico = boost::filesystem::path(yitem->iconName()).stem().c_str();
        GtkWidget *icon;
        GtkWidget *box = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 6);
        if (gtk_icon_theme_has_icon (theme, ico.c_str()))
        {
          icon = gtk_image_new_from_icon_name (ico.c_str(), GTK_ICON_SIZE_MENU);
        }
        else
        {
          icon = gtk_image_new_from_file(yitem->iconName().c_str());
        }
        GtkWidget *label = gtk_label_new (action_name.c_str());
        entry = gtk_menu_item_new ();

        gtk_container_add (GTK_CONTAINER (box), icon);
        gtk_container_add (GTK_CONTAINER (box), label);

        gtk_label_set_use_underline (GTK_LABEL (label), TRUE);
        gtk_label_set_xalign (GTK_LABEL (label), 0.0);

        gtk_container_add (GTK_CONTAINER (entry), box);
      }
      else
      {
        entry = gtk_menu_item_new_with_mnemonic (action_name.c_str());
      }
      d->menu_entry.insert(MenuEntryPair(*it, entry));


      YMGAMenuItem *menuItem = dynamic_cast<YMGAMenuItem *>(*it);
      if (menuItem)
      {
        gtk_widget_set_sensitive(entry, menuItem->enabled() ? gtk_true() : gtk_false());

        yuiDebug() << menuItem->label() << " enabled: " << menuItem->enabled() << " hidden:" << menuItem->hidden() << std::endl;
      }
      if ((*it)->hasChildren()) {
        GtkWidget *submenu = gtk_menu_new();

        gtk_menu_item_set_submenu(GTK_MENU_ITEM(entry), submenu);
        gtk_menu_shell_append (GTK_MENU_SHELL (menu), entry);
        gtk_widget_show_all(entry);

        doCreateMenu (submenu, (*it)->childrenBegin(), (*it)->childrenEnd());

      }
      else
      {
        gtk_menu_shell_append (GTK_MENU_SHELL (menu), entry);
        gtk_widget_show_all (entry);

        gulong id = g_signal_connect (G_OBJECT (entry), "activate",
                          G_CALLBACK (selected_menuitem), *it);

        d->menu_cb.insert(MenuCBPair(*it, id));
      }
      if (menuItem)
        if (menuItem->hidden())
          gtk_widget_hide(entry);
    }
  }
}


void YMGAGMenuBar::addItem(YItem* yitem)
{
  YMenuItem * item = dynamic_cast<YMenuItem *> ( yitem );
  YUI_CHECK_PTR ( item );
  yuiDebug() << item->label() << std::endl;

  // TODO icon from item
  GtkWidget *menu = gtk_menu_new();


  std::string lbl = YGUtils::mapKBAccel (item->label());
  GtkWidget *menu_entry = gtk_menu_item_new_with_mnemonic(lbl.c_str());

  d->menu_widgets.push_back(menu_entry);

  d->menu_entry.insert(MenuEntryPair(yitem, menu_entry));
  gtk_menu_item_set_submenu(GTK_MENU_ITEM(menu_entry), menu);
  gtk_widget_show(menu_entry);
  gtk_menu_shell_append (GTK_MENU_SHELL (d->menubar), menu_entry);

  YMGAMenuItem *menuItem = dynamic_cast<YMGAMenuItem *>(yitem);
  if (menuItem)
  {
    yuiDebug() << menuItem->label() << " enabled: " << menuItem->enabled() << " hidden:" << menuItem->hidden() << std::endl;
    gtk_widget_set_sensitive(menu_entry, menuItem->enabled() ? gtk_true() : gtk_false());
  }
  if (item->hasChildren())
    doCreateMenu(menu, item->childrenBegin(), item->childrenEnd());

  if (menuItem)
      if (menuItem->hidden())
        gtk_widget_hide(menu_entry);


  YMGAMenuBar::addItem(yitem);
}

void YMGAGMenuBar::enableItem(YItem* menu_item, bool enable)
{
  YMGAMenuBar::enableItem(menu_item, enable);

  auto search = d->menu_entry.find( menu_item );
  if (search != d->menu_entry.end())
  {
    GtkWidget * menu_entry = search->second;
    gtk_widget_set_sensitive(menu_entry, enable ? gtk_true() : gtk_false());
  }
  else
  {
    yuiError() << menu_item->label() << " not found" << std::endl;
  }
}

void YMGAGMenuBar::hideItem(YItem* menu_item, bool invisible)
{
  YMGAMenuBar::hideItem(menu_item, invisible);

  auto search = d->menu_entry.find( menu_item );
  if (search != d->menu_entry.end())
  {
    GtkWidget * menu_entry = search->second;
    gtk_widget_set_visible(menu_entry, invisible ? gtk_false() : gtk_true());
//     if (!invisible)
//     {
//       int min_width = this->getMinSize (YD_HORIZ);
//       int min_height = this->getMinSize (YD_VERT);
//       ygtk_adj_size_set_min(YGTK_ADJ_SIZE(getLayout()), min_width + menu_item->label().size(), min_height);
//
//       gtk_widget_queue_resize(getWidget());
//     }

  }
  else
  {
    yuiError() << menu_item->label() << " not found" << std::endl;
  }
}

void YMGAGMenuBar::deleteAllItems()
{

  for (MenuCBMap::iterator it=d->menu_cb.begin(); it!=d->menu_cb.end(); ++it)
  {
    auto search = d->menu_entry.find( it->first );
    if (search != d->menu_entry.end())
      g_signal_handler_disconnect (search->second, it->second);
  }


  for (GtkWidget *m: d->menu_widgets)
  {
    gtk_widget_destroy(m);
  }

//   for (MenuEntryMap::iterator it=d->menu_entry.begin(); it!=d->menu_entry.end(); ++it)
//   {
//     gtk_widget_destroy(it->second);
//   }

  d->menu_widgets.clear();
  d->menu_cb.clear();
  d->menu_entry.clear();

  YSelectionWidget::deleteAllItems();
}


YMGAGMenuBar::~YMGAGMenuBar()
{
  d->menu_cb.clear();
  d->menu_entry.clear();

  delete d;
}




src/YMGA_GCBTable.h
/*
  Copyright 2013 by Angelo Naselli <anaselli at linux dot it>
 
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:       YMGA_GCBTable.h

  Author:     Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YMGA_GCBTable_h
#define YMGA_GCBTable_h


#include <yui/mga/YMGA_CBTable.h>
#include <yui/gtk/ygtktreeview.h>


#include <yui/gtk/YGSelectionStore.h>
#include <yui/gtk/YGWidget.h>

/* YMGA_GTreeView is YGTreeView into gtk implementation that unfortunately is not usable
   TODO if needed by other classes move to a new header file
 */

class YMGA_GTreeView : public YGScrolledWidget, public YGSelectionStore
{
protected:
    guint m_blockTimeout;
    int markColumn;
    GtkWidget *m_count;

public:
    YMGA_GTreeView (YWidget *ywidget, YWidget *parent, const std::string &label, bool tree);

    virtual ~YMGA_GTreeView();

    inline GtkTreeView *getView()
    {
        return GTK_TREE_VIEW (getWidget());
    }
    inline GtkTreeSelection *getSelection()
    {
        return gtk_tree_view_get_selection (getView());
    }

    void addTextColumn (int iconCol, int textCol);
    void addTextColumn (const std::string &header, YAlignmentType align, int icon_col, int text_col);

    void addCheckColumn (int check_col);
    void addCheckColumn ( const std::string& header, int check_col );

    void readModel();

    void addCountWidget (YWidget *yparent);

    void syncCount();

    void focusItem (YItem *item, bool select);

    void unfocusAllItems();

    void unmarkAll();

    YItem *getFocusItem();

    virtual bool _immediateMode() {
        return true;
    }
    virtual bool _shrinkable() {
        return false;
    }
    virtual bool _recursiveSelection() {
        return false;
    }

    void setMark (GtkTreeIter *iter, YItem *yitem, gint column, bool state, bool recursive);

    void toggleMark (GtkTreePath *path, gint column);

    virtual unsigned int getMinSize (YUIDimension dim);

protected:
    static gboolean block_selected_timeout_cb (gpointer data);

    void blockSelected();

    static void block_init_cb (GtkWidget *widget, YMGA_GTreeView *pThis);

    static bool all_marked (GtkTreeModel *model, GtkTreeIter *iter, int mark_col);

    static void inconsistent_mark_cb (GtkTreeViewColumn *column,
                                      GtkCellRenderer *cell, GtkTreeModel *model, GtkTreeIter *iter, gpointer data);

    static void selection_changed_cb (GtkTreeSelection *selection, YMGA_GTreeView *pThis);

    static void activated_cb (GtkTreeView *tree_view, GtkTreePath *path,
                              GtkTreeViewColumn *column, YMGA_GTreeView* pThis);

    static void toggled_cb (GtkCellRendererToggle *renderer, gchar *path_str,
                            YMGA_GTreeView *pThis);
    static void right_click_cb (YGtkTreeView *view, gboolean outreach, YMGA_GTreeView *pThis);

};


class YMGA_GCBTable : public YMGA_CBTable, public YMGA_GTreeView
{
public:
    YMGA_GCBTable (YWidget *parent, YCBTableHeader *headers );

    void setSortable (bool sortable);

    void setCell (GtkTreeIter *iter, int column, const YTableCell *cell);

    // YGTreeView
    virtual bool _immediateMode();

    // YTable
    virtual void setKeepSorting (bool keepSorting);

    virtual void cellChanged (const YTableCell *cell);

    // YGSelectionStore
    void doAddItem (YItem *_item);

    void doSelectItem (YItem *item, bool select);

    void doDeselectAllItems();

    virtual void setItemChecked( YItem * item, int column, bool checked = true );
    
    // callbacks
    static void activateButton (YWidget *button);

    static void hack_right_click_cb (YGtkTreeView *view, gboolean outreach, YMGA_GCBTable *pThis);

    static gboolean key_press_event_cb (GtkWidget *widget, GdkEventKey *event, YMGA_GCBTable *pThis);

    static gint tree_sort_cb (GtkTreeModel *model, GtkTreeIter *a, GtkTreeIter *b, gpointer _index);

//     YGLABEL_WIDGET_IMPL (YTable)
//     YGSELECTION_WIDGET_IMPL (YTable)
    YGLABEL_WIDGET_IMPL (YMGA_CBTable)
    YGSELECTION_WIDGET_IMPL (YMGA_CBTable)
};




#endif // YQLabel_h


src/YMGAGWidgetFactory.h
/*
  Copyright 2013 by Angelo Naselli <anaselli at linux dot it>
 
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:		YMGAGWidgetFactory.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YMGAGWidgetFactory_h
#define YMGAGWidgetFactory_h


#include <yui/mga/YMGA_CBTable.h>
#include <yui/mga/YMGAWidgetExtensionFactory.h>

// #include "YMGA_CBTable.h"


using std::string;


/**
 * Concrete widget factory for mandatory widgets.
 **/
class YMGAGWidgetFactory: public YMGAWidgetFactory
{
public:

  virtual YMGA_CBTable * createCBTable ( YWidget * parent, YTableHeader * header_disown );
  virtual YMGAMenuBar  * createMenuBar ( YWidget *parent );

protected:

    friend class YGWE;

    /**
     * Constructor.
     *
     * Use YUI::widgetFactory() to get the singleton for this class.
     **/
    YMGAGWidgetFactory();

    /**
     * Destructor.
     **/
    virtual ~YMGAGWidgetFactory();

}; // class YWidgetFactory


#endif // YMGAGWidgetFactory_h


src/YMGAGWidgetFactory.cc
/*
  Copyright 2013 by Angelo Naselli <anaselli at linux dot it>
 
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGAGWidgetFactory.cc

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#define YUILogComponent "mga-gtk"
#include <yui/YUILog.h>

#include "YMGAGWidgetFactory.h"
// #include <yui/gtk/YGUI.h>
#include <yui/YUIException.h>
#include <YExternalWidgets.h>


#include <string>

#include "YMGA_GCBTable.h"
#include "YMGAGMenuBar.h"
using std::string;


YMGAGWidgetFactory::YMGAGWidgetFactory()
    : YMGAWidgetFactory()
{
    // NOP
}


YMGAGWidgetFactory::~YMGAGWidgetFactory()
{
    // NOP
}


YMGA_CBTable * YMGAGWidgetFactory::createCBTable( YWidget * parent, YTableHeader * header)
{
    YCBTableHeader *hdr = dynamic_cast<YCBTableHeader *>(header);
    YUI_CHECK_NEW(hdr);
    YMGA_GCBTable * table = new YMGA_GCBTable( parent, hdr );
    YUI_CHECK_NEW( table );

    return table;
}

YMGAMenuBar *YMGAGWidgetFactory::createMenuBar ( YWidget *parent )
{
  YMGAMenuBar *menubar = new YMGAGMenuBar(parent);
  YUI_CHECK_NEW( menubar );

  return menubar;
}
